### ISSUE TITLE
torch.compile does not work since 2.2.1 on MacOS for some models

### ISSUE THREAD

**Issue by maxbartel** [pytorch/pytorch#124497]
### 🐛 Describe the bug

The execution hangs when first calling the model to warm-up. _After_ will never be printed. This is on an Apple M3 Max. Smaller models work sometimes.

```python
import torch
from transformers import AutoModelForImageClassification


neural_network = AutoModelForImageClassification.from_pretrained("microsoft/resnet-50")
input_data = torch.randn(1, 3, 228, 228)
print(neural_network(input_data))
neural_network_c = torch.compile(neural_network, backend="inductor")
print("Before")
neural_network_c(input_data)
print("After")
```

The last version that worked was torch 2.2.0. 

### Versions

```
Collecting environment information...
PyTorch version: 2.2.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.28.4
Libc version: N/A

Python version: 3.11.8 (main, Feb  6 2024, 21:21:21) [Clang 15.0.0 (clang-1500.1.0.2.5)] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.16.0
[pip3] torch==2.2.1
[conda] Could not collect
```

Thanks for your help!

cc @ezyang @gchanan @zou3519 @kadeng @malfet @albanD @msaroufim @bdhirsh @anijain2305 @chauhang

**Comment by maxbartel** [comment]
I see the same behavior on the torch 2.3.0 RC.

**Comment by malfet** [comment]
@maxbartel can you run `pip free` and share it's output here? Also, do you mind sampling hanged pytorch process in Activity Monitor and share the results?

**Comment by anijain2305** [comment]
Can you try adding line `torch._dynamo.config.enable_cpp_guard_manager = False` in the script and see if you still see the issue? I don't have any reason to believe that it should matter, but given the timing of this issue, I want to be sure.

**Comment by maxbartel** [comment]
> @maxbartel can you run `pip free` and share it's output here? Also, do you mind sampling hanged pytorch process in Activity Monitor and share the results?

```
absl-py==2.1.0
argcomplete==3.2.3
astunparse==1.6.3
certifi==2024.2.2
cfgv==3.4.0
charset-normalizer==3.3.2
click==8.1.7
colorama==0.4.6
commitizen==3.21.3
dataclasses==0.6
decli==0.6.1
distlib==0.3.8
dm-tree==0.1.8
filelock==3.13.2
flatbuffers==24.3.25
fsspec==2024.3.1
gast==0.5.4
google-pasta==0.2.0
grpcio==1.62.1
h5py==3.10.0
huggingface-hub==0.22.0
identify==2.5.35
idna==3.6
importlib_metadata==7.1.0
iree-compiler==20240410.859
iree-runtime==20240410.859
iree-turbine==2.3.0rc20240410
Jinja2==3.1.3
julius==0.2.7
keras-nightly==3.0.4.dev2024021403
libclang==18.1.1
Markdown==3.6
markdown-it-py==3.0.0
MarkupSafe==2.1.5
mdurl==0.1.2
ml-dtypes==0.3.2
mpmath==1.3.0
namex==0.0.7
netron==7.5.6
networkx==3.2.1
nodeenv==1.8.0
numpy==1.26.4
onnx==1.16.0
opt-einsum==3.3.0
packaging==24.0
platformdirs==4.2.0
pre-commit==3.7.0
prompt-toolkit==3.0.36
protobuf==4.25.3
Pygments==2.17.2
PyYAML==6.0.1
questionary==2.0.1
regex==2023.12.25
requests==2.31.0
rich==13.7.1
# Editable Git install with no remote (roofbench==0.0.post1.dev1+g2cb50cf.d20240325)
-e /Users/bartel/Documents/roofbench
safetensors==0.4.2
six==1.16.0
sympy==1.12
tb-nightly==2.16.0a20240212
tensorboard-data-server==0.7.2
tensorflow-io-gcs-filesystem==0.36.0
termcolor==2.4.0
tf-nightly==2.16.0.dev20240207
tf_keras-nightly==2.16.0.dev2024021410
tokenizers==0.15.2
tomlkit==0.12.4
torch==2.3.0
tqdm==4.66.2
transformers==4.39.1
typer==0.11.0
typing_extensions==4.10.0
urllib3==2.2.1
virtualenv==20.25.1
wcwidth==0.2.13
Werkzeug==3.0.1
wrapt==1.14.1
zipp==3.18.1
```

and here is a gist with the sampling https://gist.github.com/maxbartel/f03722d324089b9d8c2605dbd8ac1433






> Can you try adding line `torch._dynamo.config.enable_cpp_guard_manager = False` in the script and see if you still see the issue? I don't have any reason to believe that it should matter, but given the timing of this issue, I want to be sure.

I tried, but looking at https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/config.py it seems like none of the releases have this option. :confused: I tried the nightly build where this option is available and it doesn't change the behavior. The program still hangs...

**Comment by malfet** [comment]
Thank you for the spindump. It confirms my theory that it deadlocks because two libomp.dylibs are loaded into the runtime.
From spindump:
```
0x110b2fff7 +libomp.dylib (0) <F53B1E01-AF16-30FC-8690-F7B131EB6CE5> /Users/*/Documents/*/libomp.dylib
...
0x32e1bfffb +libomp.dylib (0) <BF269055-21D9-3FED-B014-E72F141706F3> /opt/homebrew/*/libomp.dylib
```

I suspect first one is bundled with PyTorch and 2nd one comes from home-brew, that compiler is trying to load.
We should put a warning against that/configure compile to use just one. In the meantime @maxbartel do you mind uninstalling one from home-brew to make sure it solves the problem?

**Comment by maxbartel** [comment]
This solved the problem! 🎉

I think you are right, compile should decide which one to use and stick to it. A warning is not too useful if `libomp` is a dependency for something else.

Thanks for your help!

**Comment by malfet** [comment]
@anijain2305 what would be the debug option to dump the compiler invocation commands? As installing libomp from homebrew did not result in a hang for me, guess because compiler picks libomp bundled with torch

**Comment by masnesral** [comment]
@malfet we're doing some maintenance in the triage meeting; I hope it's not presumptuous to assign the issue to you (seems like you're handling it?)

**Comment by aisven** [comment]
For PyTorch users with Apple machines with Intel CPUs,

will there be fixes of `torch.compile` released for PyTorch 2.2.x like 2.2.3 ?

I am really interested in the `torch.compile` capability, but I am afraid that as a university student I only have two MacBook Pro laptops that both still have Intel CPUs, and unfortunately due to the sad affair https://github.com/pytorch/pytorch/issues/114602 I cannot just upgrade to a PyTorch 2.3.x version. Please advise.

**Comment by redradist** [comment]
Is there any update on this issue ?

**Comment by atalman** [comment]
Removing from 2.3.1 Milestone. - 5/27 Cherry-pick post deadline (End of day 5PM PST)
